File: test/atcoder-past202012-n.test.cpp
#define PROBLEM "https://atcoder.jp/contests/past202012-open/tasks/past202012_n"

#include "../cpp/segtree.hpp"
#include <iostream>

int main() {
    int n, q; std::cin >> n >> q;
    std::vector<std::pair<int, int>> lr(n-1);
    for(int i = 0; i < n-1; i++) {
        int l, r; std::cin >> l >> r;
        lr[i] = {l, r};
    }
    SegTree seg(lr, [](const std::pair<int, int>& a, const std::pair<int, int>& b) {
            return std::pair<int, int>(std::max(a.first, b.first), std::min(a.second, b.second));
            }, std::pair<int,int>(0, 1e9));
    while(q--) {
        int a, b; std::cin >> a >> b; b--;
        int r = seg.max_right(b, [&](const std::pair<int, int>& x) { return x.first <= a && a <= x.second; });
        int l = seg.min_left(b, [&](const std::pair<int, int>& x) { return x.first <= a && a <= x.second; });
        std::cout << r - l + 1 << std::endl;
    }
}

File: test/atcoder-past202012-n.test.py
# verification-helper: PROBLEM https://atcoder.jp/contests/past202012-open/tasks/past202012_n
from py.segtree import SegTree

N, Q = map(int, input().split())
lr = [tuple(map(int, input().split())) for i in range(N-1)]
seg = SegTree(lr, lambda a, b: (max(a[0], b[0]), min(a[1], b[1])), (0, 10**9))
for i in range(Q):
    a, b = map(int, input().split())
    b -= 1
    r = seg.max_right(b, lambda x: x[0] <= a <= x[1])
    l = seg.min_left(b, lambda x: x[0] <= a <= x[1])
    print(r-l+1)
